File: app/src/main/java/cz/johnyapps/cheers/adapters/SelectableAdapter.java
package cz.johnyapps.cheers.adapters;

import android.content.Context;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.res.ResourcesCompat;
import androidx.recyclerview.widget.RecyclerView;

import cz.johnyapps.cheers.R;

public abstract class SelectableAdapter<VIEW_HOLDER extends SelectableAdapter<VIEW_HOLDER, ITEM, DATA>.SelectableViewHolder, ITEM, DATA> extends BaseAdapter<VIEW_HOLDER, DATA> {
    @Nullable
    private SelectedItem<ITEM> selectedItem;
    @Nullable
    private OnSelectListener<ITEM> onSelectListener;
    private boolean allowSelection = true;

    public SelectableAdapter(@NonNull Context context) {
        super(context);
    }

    @Override
    public void onBindViewHolder(@NonNull VIEW_HOLDER holder, int position) {
        boolean selected = selectedItem != null && selectedItem.getPosition() == position;

        if (selected) {
            holder.itemView.setForeground(ResourcesCompat.getDrawable(getContext().getResources(),
                    R.drawable.selected_item_background,
                    getContext().getTheme()));
        } else {
            holder.itemView.setForeground(null);
        }

        onBindViewHolder(holder, position, selected);
    }

    public abstract void onBindViewHolder(@NonNull VIEW_HOLDER holder, int position, boolean selected);

    @Override
    public void update(@Nullable DATA data) {
        selectedItem = null;

        if (onSelectListener != null) {
            onSelectListener.onSelect(null);
        }

        super.update(data);
    }

    @NonNull
    public abstract ITEM getItem(int position);

    public static class SelectedItem<ITEM> {
        @NonNull
        private final ITEM selectedItem;
        private final int position;

        public SelectedItem(@NonNull ITEM selectedItem, int position) {
            this.selectedItem = selectedItem;
            this.position = position;
        }

        @NonNull
        public ITEM getSelectedItem() {
            return selectedItem;
        }

        public int getPosition() {
            return position;
        }
    }

    public void selectPosition(int pos) {
        if (!allowSelection) {
            return;
        }

        SelectedItem<ITEM> oldItem = selectedItem;

        if (oldItem != null && oldItem.getPosition() == pos || pos < 0) {
            selectedItem = null;
        } else {
            selectedItem = new SelectedItem<>(getItem(pos), pos);
        }

        if (oldItem != null) {
            notifyItemChanged(oldItem.getPosition());
        }

        if (selectedItem != null && (oldItem == null || oldItem.getPosition() != selectedItem.getPosition())) {
            notifyItemChanged(selectedItem.getPosition());
        }

        if (onSelectListener != null) {
            onSelectListener.onSelect(selectedItem == null ? null : selectedItem.getSelectedItem());
        }
    }

    public void setAllowSelection(boolean allowSelection) {
        this.allowSelection = allowSelection;
    }

    public boolean isAllowSelection() {
        return allowSelection;
    }

    public class SelectableViewHolder extends RecyclerView.ViewHolder {
        public SelectableViewHolder(@NonNull View root) {
            super(root);
            root.setOnLongClickListener(v -> {
                selectPosition(getAdapterPosition());
                return false;
            });
            root.setOnClickListener(v -> {
                if (selectedItem != null) {
                    selectPosition(getAdapterPosition());
                }
            });
        }
    }

    public void setOnSelectListener(@Nullable OnSelectListener<ITEM> onSelectListener) {
        this.onSelectListener = onSelectListener;
    }

    public interface OnSelectListener<ITEM> {
        void onSelect(@Nullable ITEM selectedItem);
    }
}

File: app/src/main/java/cz/johnyapps/cheers/adapters/SelectableAdapter.kt
package cz.johnyapps.cheers.adapters

import android.view.View
import androidx.core.content.res.ResourcesCompat
import androidx.recyclerview.widget.DiffUtil
import cz.johnyapps.cheers.R

import java.util.*
import kotlin.collections.Collection
import kotlin.collections.HashMap
import kotlin.collections.isNotEmpty
import kotlin.collections.set
import kotlin.collections.toTypedArray

abstract class SelectableAdapter<T, VH: SelectableAdapter<T, VH>.ViewHolder>(itemCallback: DiffUtil.ItemCallback<T>): BaseAdapter<T, VH>(itemCallback) {
    private val selectedItems = HashMap<Int, T>()
    private var multiSelection = true
    private var allowSelection = true
    private var canceledSelectionThisClick = false
    var onSelectListener: OnSelectListener<T>? = null

    fun isSelecting(): Boolean {
        return selectedItems.isNotEmpty()
    }

    open fun isSelected(position: Int): Boolean {
        return selectedItems[position] != null
    }

    open fun selectPosition(pos: Int) {
        if (!allowSelection) {
            return
        }

        val oldItemPos = if (selectedItems.keys.isEmpty()) -1 else selectedItems.keys.toTypedArray()[0]

        val wasAlreadySelected: Boolean = isSelected(pos)

        if (wasAlreadySelected || pos < 0) {
            selectedItems.remove(pos)
        } else {
            if (!multiSelection && oldItemPos >= 0) {
                selectedItems.remove(oldItemPos)
                notifyItemChanged(oldItemPos)
            }

            val selectedItem: T = getItem(pos)
            selectedItems[pos] = selectedItem
        }

        notifyItemChanged(pos)

        onSelectListener?.onSelect(selectedItems.values)
    }

    open fun cancelSelection() {
        val positions: Collection<Int> = ArrayList(selectedItems.keys)
        for (pos in positions) {
            selectPosition(pos)
        }
    }

    open fun selectAllOrCancelSelection() {
        var allSelected = true

        for (i in 0 until itemCount) {
            if (!isSelected(i)) {
                allSelected = false
                selectPosition(i)
            }
        }

        if (allSelected) {
            cancelSelection()
        }
    }

    override fun onBindViewHolder(holder: VH, position: Int) {
        val selected = isSelected(position)

        holder.itemView.setOnLongClickListener {
            selectPosition(position)
            true
        }

        if (selected) {
            holder.itemView.foreground = ResourcesCompat.getDrawable(
                getContext(holder).resources,
                R.drawable.selected_item_background,
                getContext(holder).theme
            )
        } else {
            holder.itemView.foreground = null
        }

        onBindViewHolder(holder, position, selected)
    }

    abstract fun onBindViewHolder(holder: VH, position: Int, selected: Boolean)

    open fun canceledSelectionThisClick(): Boolean {
        return canceledSelectionThisClick
    }

    open fun setAllowSelection(allowSelection: Boolean) {
        this.allowSelection = allowSelection
    }

    open fun isAllowSelection(): Boolean {
        return allowSelection
    }

    open fun setMultiSelection(multiSelection: Boolean) {
        this.multiSelection = multiSelection
    }

    open inner class ViewHolder(itemView: View): BaseAdapter<T, VH>.ViewHolder(itemView) {
        init {
            itemView.setOnLongClickListener {
                selectPosition(adapterPosition)
                false
            }

            addRootOnClickListener(object : RootOnClickListener {
                override fun onClick(view: View) {
                    canceledSelectionThisClick = if (isSelecting()) {
                        selectPosition(adapterPosition)
                        selectedItems.isEmpty()
                    } else {
                        false
                    }
                }
            })
        }
    }

    interface OnSelectListener<T> {
        fun onSelect(selectedItems: Collection<T>?)
    }
}